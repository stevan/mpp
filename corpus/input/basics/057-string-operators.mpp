# String concatenation
my $concat = "Hello" . " " . "World";
my $concat2 = $first . $last;

# String repetition
my $repeat = "abc" x 3;
my $dashes = "-" x 10;
my $pattern = "01" x $count;

# Combined string operations
my $header = "=" x 40 . "\n";
my $formatted = $prefix . ($char x $width) . $suffix;

# String concatenation assignment
$message .= "!";
$result .= $line . "\n";

# Complex string expressions
my $output = "Name: " . $name . " (ID: " . $id . ")";
my $border = "+" . ("-" x ($width - 2)) . "+";
my $repeated_pattern = ($pattern x 3) . "\n";