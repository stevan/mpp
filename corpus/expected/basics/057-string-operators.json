[
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$concat"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "String",
          "value": "\"Hello\""
        },
        "right": {
          "type": "String",
          "value": "\" \""
        }
      },
      "right": {
        "type": "String",
        "value": "\"World\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$concat2"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "Variable",
        "name": "$first"
      },
      "right": {
        "type": "Variable",
        "name": "$last"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$repeat"
    },
    "initializer": {
      "type": "String",
      "value": "\"abc\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$dashes"
    },
    "initializer": {
      "type": "String",
      "value": "\"-\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$pattern"
    },
    "initializer": {
      "type": "String",
      "value": "\"01\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$header"
    },
    "initializer": {
      "type": "String",
      "value": "\"=\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$formatted"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "Variable",
          "name": "$prefix"
        },
        "right": {
          "type": "Variable",
          "name": "$char"
        }
      },
      "right": {
        "type": "Variable",
        "name": "$suffix"
      }
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$message"
    },
    "operator": ".=",
    "right": {
      "type": "String",
      "value": "\"!\""
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$result"
    },
    "operator": ".=",
    "right": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "Variable",
        "name": "$line"
      },
      "right": {
        "type": "String",
        "value": "\"\\n\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$output"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "BinaryOp",
          "operator": ".",
          "left": {
            "type": "BinaryOp",
            "operator": ".",
            "left": {
              "type": "String",
              "value": "\"Name: \""
            },
            "right": {
              "type": "Variable",
              "name": "$name"
            }
          },
          "right": {
            "type": "String",
            "value": "\" (ID: \""
          }
        },
        "right": {
          "type": "Variable",
          "name": "$id"
        }
      },
      "right": {
        "type": "String",
        "value": "\")\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$border"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "String",
          "value": "\"+\""
        },
        "right": {
          "type": "String",
          "value": "\"-\""
        }
      },
      "right": {
        "type": "String",
        "value": "\"+\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$repeated_pattern"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "Variable",
        "name": "$pattern"
      },
      "right": {
        "type": "String",
        "value": "\"\\n\""
      }
    }
  }
]